warp10/src/main/java/io/warp10/script/ext/http/HTTP.java

```diff
@@ -0,0 +1,278 @@
+import java.util.concurrent.locks.ReentrantLock;
+
+/**
+ * Apply an HTTP method over an url
+ */
+public class HTTP extends FormattedWarpScriptFunction {
+
+  //
+  // Arguments
+  //
+
+  public static final String METHOD = "method";
+  public static final String URL = "url";
+  public static final String HEADER = "header";
+  public static final String BODY = "body";
+  public static final String RESPONSE = "response";
+
+  private final Arguments args;
+  private final Arguments output;
+
+  protected Arguments getArguments() {
+    return args;
+  }
+  protected Arguments getOutput() {
+    return output;
+  }
+
+  //
+  // Control
+  //
+
+  private final ReentrantLock stackCountersLock = new ReentrantLock();
+  private final WebAccessController webAccessController;
+
+  public HTTP(String name) {
+    super(name);
+
+    String patternConf = WarpConfig.getProperty(HttpWarpScriptExtension.WARPSCRIPT_HTTP_HOST_PATTERNS);
+
+    // If not defined, use already existing StandaloneWebCallService webAccessController which uses Configuration.WEBCALL_HOST_PATTERNS
+    if (null == patternConf) {
+      webAccessController = StandaloneWebCallService.getWebAccessController();
+    } else {
+      webAccessController = new WebAccessController(patternConf);
+    }
+
+    getDocstring().append("Apply an HTTP method over an url and fetch response.");
+
+    args = new ArgumentsBuilder()
+      .addArgument(String.class, METHOD, "The http method.")
+      .addArgument(String.class, URL, "The URL to send the GET request to. Must begin with http:// or https://.")
+      .addOptionalArgument(Map.class, HEADER, "An optional header.", new HashMap<>())
+      .addOptionalArgument(String.class, BODY, "An optional body", "")
+      .build();
+
+    output = new ArgumentsBuilder()
+      .addArgument(List.class, RESPONSE, "A list containing, for each URL, a 4-element list. These 4-element lists contain, in this order, a LONG status code, a STRING status message or an empty STRING if not available, a MAP of headers and a STRING representing a bytes array encoded as base 64.")
+      .build();
+  }
+
+  @Override
+  public WarpScriptStack apply(Map<String, Object> formattedArgs, WarpScriptStack stack) throws WarpScriptException {
+
+    //
+    // Check capability if set in configuration, or else check stack authentication
+    //
+
+    Properties props = WarpConfig.getProperties();
+
+    if (props.containsKey(HttpWarpScriptExtension.HTTP_CAPABILITY)) {
+      if (stack.getAttribute(WarpScriptStack.CAPABILITIES_ATTR) instanceof Capabilities) {
+        Capabilities capabilities = (Capabilities) stack.getAttribute(WarpScriptStack.CAPABILITIES_ATTR);
+
+        if (!(capabilities.containsKey(props.getProperty(HttpWarpScriptExtension.HTTP_CAPABILITY)))) {
+          throw new WarpScriptException("Capability " + HttpWarpScriptExtension.HTTP_CAPABILITY + " is required by function " + getName());
+        }
+      }
+
+    } else if (!stack.isAuthenticated()) {
```

Comment:
The authorization type should be configurable, either none, i.e. the function can be used by anyone if the extension is loaded, authenticated, AUTHENTICATE must have been called, and capability to inspect a capability.

Reply:
I was thinking of enforcing authentication in each case (capability or not), since it allows for checking if another thread in the jvm has a stack that is authenticated with the same token and also try to use HTTP, so that we can prevent nested call of HTTP and network loops.